### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by jkarni**
Check person is authorized for changes. E.g., only admin/owner in project, only author can retract a diff, only admin can accept or reject a diff, only author or owner can change issues.

### CODE REVIEW

File: rad/monadic/project.rad [hunk 1]
```diff
@@ -87,4 +90,12 @@
...
   [ (lookup :id (first-rsm-of-type! machine :rad-diff)) ==> "b" ]
   [ (lookup :id (first-rsm-of-type! machine :rad-repo)) ==> "c" ]
+  ;; authorization
+  [ (do
+       (set-fake-keys! non-owner-keys)
+       (catch 'validation-failure
+              (do (add-diff-rsm! machine "d")
+                  :valid)
+              (fn [_] :not-valid))) ==> :not-valid
+  ]
```
Comment (MeBrei) [inline rad/monadic/project.rad]: Maybe add some tests for state changes?

File: rad/prelude/auth.rad [hunk 2]
```diff
@@ -0,0 +1,85 @@
+{:module 'prelude/auth
+ :doc "Functions for easily handling authorization on chains."
```
Comment (MeBrei) [inline rad/prelude/auth.rad]: `on machines.` ?

File: rad/prelude/auth.rad [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  A common pattern is to set up a machine, and only then set the owner of the
+  machine. Since setting the owner is something only owners can do, we provide
+  a way for *initializing* the authentication mechanism. Prior to
+  initialization, `(auth [:allowed <group> <person>])` always succeeds.
+  After `(auth [:initialize])` is called, it returns whatever
+  `(auth [:is-in <group> <person>])` returns.
+"
+  (fn [set-names]
+    (def groups (ref (dict-from-seq (map (fn [k] [k (ref set/empty)]) set-names))))
```
Comment (jameshaydon) [inline rad/prelude/auth.rad]: Why do we need refs at the keys?

File: rad/monadic/project-remote.rad [hunk 4]
```diff
@@ -28,6 +29,18 @@
 
 (def noncer (make-noncer))
 
+(def auth (make-simple-auth [:admin]))
+
+(def add-admin
+  "Add an admin to the diff machine. If the auth is initialized, only an admin
```
Comment (jameshaydon) [inline rad/monadic/project-remote.rad]: ~~diff machine~~ project machine

File: rad/monadic/project-remote.rad [hunk 5]
```diff
@@ -28,6 +29,18 @@
...
+  is allowed to do so. Otherwise, this initializes the auth."
+  (fn [pub-key-msg]
+    (def pub-key (lookup :key pub-key-msg))
+    (def author-key (lookup :author pub-key-msg))
+    ((auth [:allowed :admin]) pub-key)
```
Comment (jameshaydon) [inline rad/monadic/project-remote.rad]: Quite cute!

File: rad/monadic/issue-remote.rad [hunk 6]
```diff
@@ -27,6 +28,18 @@
 
 (def noncer (make-noncer))
 
+(def auth (make-simple-auth [:admin]))
+
+(def add-admin
```
Comment (jameshaydon) [inline rad/monadic/issue-remote.rad]: Can we add this fn to the `auth` module.

File: rad/monadic/diff-remote.rad [hunk 7]
```diff
@@ -103,11 +116,25 @@
 
 (def validator/diff-edit-input
   "An input to edit an diff."
-  (validator/and
-   [validator/input
-    validator/time-modified
-    validator/diff-reference
-    (validator/contains-only (edit-keys (keys diff-keys)))]))
+  (fn [ed]
+    (def always-valid (fn [x] x))
```
Comment (jameshaydon) [inline rad/monadic/diff-remote.rad]: We can add this to `validation`.

File: rad/prelude/auth.rad (hunk 3 above)
Comment (jkarni) [reply rad/prelude/auth.rad]: Don't need it, but feels slightly easier to modify.

File: bin/rad-diff [hunk 8]
```diff
@@ -260,7 +260,16 @@
         (do
           (def edit-state-with-feedback!
             (fn [message]
-              (simple-edit-diff! machine n {:state new-state})
+              (catch 'daemon-error
```
[the remark this answers is not in the document]
Comment (jkarni) [reply bin/rad-diff]: Created a trello issue for that.

**Comment by jkarni** [comment]
@geigerzaehler mind taking a look at the second commit? changed a few things in cloudbuild (most importantly, to larger instances) that I think are worth being aware of.

**Pull request merged**